File: baekjoon/Java/1182_부분수열의 합.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int ans=0;
    static int N,S;
    static int[] lst;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        S = Integer.parseInt(st.nextToken());
        lst = new int[N];
        st = new StringTokenizer(br.readLine());
        for(int i=0; i<N; i++){
            lst[i] = Integer.parseInt(st.nextToken());
        }


        dfs(0,0,0);
        System.out.println(ans);
    }
    public static void dfs(int depth,int sum,int cnt){
        if(depth==N){
            if(sum==S && cnt>0){
                ans+=1;
            }
            return;
        }
        dfs(depth+1,sum+lst[depth],cnt+1);
        dfs(depth+1,sum,cnt);
    }
}

File: baekjoon/Python/1182_부분수열의 합.py
N, S = map(int, input().split())
lst = list(map(int, input().split()))
sum = 0
ans = 0


def dfs(depth, sum, cnt):
    global ans
    # 종료조건 + 정답처리
    if depth == N:
        if sum == S and cnt > 0:
            ans += 1
        return

    dfs(depth + 1, sum + lst[depth], cnt + 1)  # 포함하는 경우
    dfs(depth + 1, sum, cnt)  # 포함하지 않는 경우


dfs(0, 0, 0)
print(ans)
